v0.1.2: new core

cathaypacific8747 opened the pull request:
The current core is quite cumbersome to use:
```py
async with FR24() as fr24:
    fl = fr24.flight_list("id")
    fl.data.load() # read flight_list/{id}.parquet
    response = await fl.api.fetch() # optional parameters goes here, e.g. page, limit
    fl.data.add(response) # json -> pyarrow
    fl.data.save() # write flight_list/{id}.parquet
```

It is unclear where data is actually stored. It would be ideal to have something like:
```py
async with FR24() as fr24:
    await fr24.login()
    data_old = fr24.flight_list.load("id") # reads flight_list/{id}.parquet
    # FlightListArrow { ctx: FlightListContext, table: pyarrow.Table }

    data_new = (
        await fr24.flight_list.fetch("id") # FlightListAPIResp { ctx: {"id": "{id}", "page": "..."}, response: json }
    ).to_arrow() # FlightListArrow { ctx: ..., table: pyarrow.Table }
    data = data_old.concat(data_new)
    data.save() # calls self.table.write_parquet("flight_list/{id}.parquet")
```
- make logins explicit instead of trying to login in `__aenter__`
- wrap around the raw/processed data with composition

Things to implement for the new core:
- `.flight_list.fetch_all()`
- `.playback.*`
- `.livefeed.*`
- `.find.*`
- store request context in arrow table `.schema.metadata`

xoolive commented on .github/workflows/docs.yml:
```diff
@@ -2,10 +2,10 @@ name: build docs
+  # pull_request_target:
+  # push:
+  #   branches:
+  #     - master
```
Commenting `pull_request_target` makes sense, but I recommend the one on push master

cathaypacific8747 commented:
The rewrite is now complete, the core now supports three major services (`livefeed`, `playback` and `flight_list`):
```py
async def main() -> None:
    async with FR24() as fr24:
        # download livefeed of entire world
        response = await fr24.livefeed.fetch() # {ctx: {...}, data: list[protobuf]}
        datac = response.to_arrow() # {ctx: {...}, data: pa.Table}
        datac.save() # saves to ~/.cache/fr24/livefeed/{timestamp}.parquet, determined from .ctx
        datac_disk = fr24.livefeed.load(timestamp) # reads data from cache and builds ctx

        # download and save trajectory for one flight
        response = await fr24.playback.fetch(0x2FB3041).to_arrow().save()
        datac_disk = fr24.playback.load(0x2FB3041)
        print(datac_disk.metadata) # accesses metadata stored in arrow schema

        # iteratively upsert pages of flight list to an existing arrow table
        data = fr24.flight_list.load(reg="B-HPB")
        async for response in fr24.flight_list.fetch_all(reg="B-HPB"):
            data_new = response.to_arrow()
            data.concat(data_new, inplace=True)
            data.save()

await main()
```
Documentation and tests are also updated for the new core. Some details:
- low level functions are left mostly untouched, except:
  - `login_with_token_subscription_key`: checks if the token is expired by parsing the JWT claim
  - `livefeed_flightdata_dict`: added missing squawk field
  - `find`: parameterised `AsyncClient` and modified parts of the TUI to use the app state's one instead
- switched to `httpx.AsyncClient(http2=True)` by default to avoid 464/525
- CLI now supports all three services
- previously, protobuf definitions are heavily nested - they have been rewritten to their match the official names and structure as far as possible
- fr24 seems to have recently added new gRPC endpoints (e.g. `flight_list`) - their definitions are partially added and we can possibly use them in the future

xoolive commented:
Hi @cathaypacific8747 will do

xoolive commented on docs/usage/cli.md:
````diff
@@ -68,6 +71,81 @@
+    
+        ```command
+        $ fr24 flight-list --reg B-HPB
+        Success: Saved 10 rows (1290 bytes) to /home/user/.cache/fr24/flight_list/reg/B-HPB.parquet.
````
Honestly... it looks weird to me from a user experience perspective that a tool just caches a result without returning anything. How about printing the flight list to the terminal (as a csv for instance?) in a way it can be piped or something?

cathaypacific8747 replied:
Yeah, I'll probably make it:
- output the pd.DataFrame repr by default
- `-o`, `--output` to save to a specific directory/filename,
- `-c`, `--cache` to save to the default cache directory

xoolive replied:
I would keep the cache on by default.
`-o` would remove the stdout output? 
I wouldn't go for some pretty repr on stdout as it is hard to pipe to more tools?

cathaypacific8747 replied:
Sure, maybe:
- `fr24 feed` saves to cache and shows a "preview" of the saved data via `pd.Dataframe` repr
- `fr24 feed --format parquet|csv|json -o /path/to/file` writes to file, no dataframe preview is shown, just how many bytes/records are written
  - `/path/to/file` can be `-` (stdout) and piped to tools, number of bytes/records would then be written to stderr (similar to curl)

cathaypacific8747 replied:
I would also like to clarify my motivation behind the CLI and more broadly the "core". Usually, when I have to analyse a particular flight, it would involve:
- `fr24.core.flight_list.fetch()` to grab the fr24 flight id, then
- `fr24.core.playback.fetch(flight_id)` to grab the trajectory
- parquet are saved to hardcoded folders
- downstream analyses can then use `.load()` to avoid expensive calls

My intention with the CLI was more of a "download manager" that works with the parquet files. Maybe the "cache" should be more appropriately named "database" 😅

cathaypacific8747 replied:
Implemented in https://github.com/cathaypacific8747/fr24/pull/24/commits/e19e5ba1d0715f9670aa703227c71b6c1bfc0c4f, CLI docs: https://bf3fe814.flightradar24.pages.dev/usage/cli

xoolive commented on docs/usage/cli.md:
```diff
@@ -68,6 +71,81 @@
+        $ fr24 flight-list --reg B-HPB
+        Success: Saved 10 rows (1290 bytes) to /home/user/.cache/fr24/flight_list/reg/B-HPB.parquet.
+        
+        $ duckdb -c "describe select * from '/home/user/.cache/fr24/flight_list/reg/B-HPB.parquet';"
```
That's very nice though, keep it!

xoolive commented on docs/usage/examples.md:
````diff
@@ -14,31 +14,35 @@ You can find even more usage examples under [`tests/`](https://github.com/cathay
     ```
 
 #### Paginate all pages
 Queries for next page as long as user doesn't enter `x`, or if there are no pages left.
````
Maybe I have to read everything in context, but in the diff, it's hard to make sense of it.

xoolive replied:
More generally, do you have the doc in this branch rendered somewhere?

cathaypacific8747 replied:
Sure, just configured cloudflare pages for all branches: https://c9384447.flightradar24.pages.dev/

xoolive commented on src/fr24/history.py:
```diff
@@ -287,10 +288,11 @@ def playback_track_ems_dict(point: TrackData) -> PlaybackTrackEMSRecord | None:
     """
     Transform each point in the flight track to a pandas DataFrame.
+
+    Note: EMS and metadata will not be serialised.
```
😢

cathaypacific8747 replied:
Fixed in https://github.com/cathaypacific8747/fr24/pull/24/commits/3edf4477455e432cc4ad89727912f63012cd16c3

xoolive commented on src/fr24/livefeed.py:
```diff
@@ -159,23 +165,25 @@ def livefeed_response_parse(data: bytes) -> LiveFeedResponse:
-        "ground_speed": lfr.ground_speed,
+        "latitude": lfr.lat,
+        "longitude": lfr.lon,
+        "heading": lfr.track,  # most notably, this might cause confusion
```
Please don't add "heading" to the confusion 😭:
ADS-B contains the true track angle; you may have information about the heading angle in some EMS entries.

See [here](https://aviationbook.netlify.app/chapters/data_sources/flight-tracking.html#ads-b) (warning box) for details

cathaypacific8747 replied:
This was fixed in https://github.com/cathaypacific8747/fr24/pull/24/commits/345b8cdeb99e25e3410f80a9aeda255597e048ed

xoolive commented on src/fr24/tui/tui.py:
```diff
@@ -70,7 +70,7 @@ class FR24(App[None]):
     def compose(self) -> ComposeResult:
         self.auth: Authentication | None = None
-        self.client = httpx.AsyncClient()
+        self.client = httpx.AsyncClient(http2=True)
```
I'm not sure why you need that option?

cathaypacific8747 replied:
This traces back to this issue: https://github.com/cathaypacific8747/fr24/issues/23#issuecomment-2125624974 - fr24's servers sometimes returns 464 errors if HTTP/1 is used. HTTP/2 was enabled by default to keep it consistent with requests made by the browser

xoolive replied:
OK all clear, would deserve a comment in the code 😉

xoolive reviewed (commented):
Also tried the text user interface, and it's still working 🥳
